content/blog/kubernetes-monitoring-using-prometheus-and-grafana-in-hpe-greenlake-for-private-cloud-enterprise.md

```diff
@@ -0,0 +1,419 @@
+
+```markdown
+$ cat main.tf 
+terraform {
+  required_providers {
+    hpegl = {
+      source  = "hpe/hpegl"
+      version = ">= 0.2.2"
+    }
+  }
+}
+
+provider "hpegl" {
+  caas {
+    api_url = var.api_url
+  }
+}
+
+data "hpegl_caas_cluster" "iaccluster" {
+  name     = var.cluster_name
+  space_id = var.space_id
+}
+
+provider "helm" {
+  kubernetes {
+       host     = yamldecode(base64decode(data.hpegl_caas_cluster.iaccluster.kubeconfig)).clusters[0].cluster.server
+       token    = yamldecode(base64decode(data.hpegl_caas_cluster.iaccluster.kubeconfig)).users[0].user.token
+       insecure = true
+  }
+}
+
+resource "helm_release" "prometheus-stack" {
+   name = "prometheus-stack"
+   repository = "https://prometheus-community.github.io/helm-charts"
+   chart = "prometheus"
+   version = "23.0.0"
+   namespace = "monitoring"
+   create_namespace = true
+
+   set {
+    name  = "server.service.type"
+    value = "NodePort"
+  }
+
+   set {
+    name  = "prometheus-node-exporter.hostRootFsMount.enabled"
+    value = "false"
+  }
+
+   set {
+    name  = "prometheus-node-exporter.hostNetwork"
+    value = "false"
+  }
+
+   set {
+    name  = "prometheus-node-exporter.hostPID"
+    value = "false"
+  }
+}
+
+resource "helm_release" "grafana-dashboard" {
+   name = "grafana-dashboard"
+   repository = "https://grafana.github.io/helm-charts"
+   chart = "grafana"
+   version = "6.57.4"
+   namespace = "monitoring"
+   create_namespace = true
+
+   set {
+    name  = "service.type"
+    value = "NodePort"
+  }
+
+   set {
+    name  = "persistence.enabled"
+    value = "true"
+  }
+}
+```
+
+There are a few things worth noting in above Terraform config file:
+
+<style> li { font-size: 100%; line-height: 23px; max-width: none; } </style>
+
+* In Grafana, the persistence by default is disabled. In case Grafana Pod gets terminated for some reason, you will lose all your data. In production deployment, such as HPE GreenLake for Private Cloud Enterprise: Containers, this needs to be enabled, by setting *persistence.enabled* as *true*, to prevent any data lose.  
+* In Prometheus, the *DaemonSet* deployment of the node exporter is trying to mount the *hostPath* volume to the container root “/”, which violates against one deployed OPA (Open Policy Agent) policy to the K8s cluster for filesystem (FS) mount protections. The DaemonSet deployment will never be ready, keep showing the warning events as *Warning  FailedCreate daemonset-controller  Error creating: admission webhook "soft-validate.hpecp.hpe.com" denied the request: Hostpath ("/") referenced in volume is not valid for this namespace because of FS Mount protections.*. You need to disable the *hostRootFsMount*, together with *hostNetwork* and *hostPID*, to comply with the security policy in the cluster.
```

Comment:
Change "is trying" to simply "tries", so it reads "...the node exporter tries to mount..."

Change "which violates against one deployed OPA" to "which violates the deployed OPA"

Modify the next sentence to read "As such, the DaemonSet deployment will never be ready and instead keeps showing warning events as..."